File: leetcode/container_with_most_water.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    int maxArea(vector<int>& a) {
        int i = 0;
        int j = a.size() - 1;
        int best = INT_MIN;
        while (i != j) {
            best = max( min(a[i], a[j]) * (j - i), best);

            if (a[i] < a[j] || a[i] == a[j]) {
                i++;
            } else {
                // a[i] > a[j]
                j--;
            }
        }

        return best;
    }
};

int main() {
    vector<int> a = {1, 2, 3, 4, 1, 1, 1, 1, 2, 3, 4};

    Solution s;
    cout << s.maxArea(a) << endl;

    return 0;
}

File: leetcode/neetcode_roadmap/container_with_most_water.py
class Solution:
    def maxArea(self, heights: list[int]) -> int:
        i = 0
        j = len(heights) - 1
        best = 0
        while i < j:
            area = (j-i) * min(heights[i], heights[j])
            best = max(best, area)

            if heights[i] < heights[j]:
                i += 1
            elif heights[i] > heights[j]:
                j -= 1
            else:
                i += 1
                j -= 1

        return best


S = Solution()
height = [1,7,2,5,4,7,3,6]
print(S.maxArea(height))
